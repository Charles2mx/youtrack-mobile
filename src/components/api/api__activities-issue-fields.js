/* @flow */
import ApiHelper from './api__helper';
import IssueFields from './api__issue-fields';

const toField = ApiHelper.toField;

const ISSUE_ACTIVITIES_EVENT_BASE = toField([
  '$type',
  'id',
  'name',
  'text',
  'color(id)',
]);

const ISSUE_ATTACHMENT_FIELDS = toField([
  'url',
  'mimeType',
  'removed',
  'thumbnailURL',
  'size',
  'created',
  {
    comment: ['id']
  }
]);

const ISSUE_PROJECT_FIELDS = toField([
  'shortName'
]);


const ISSUE_ACTIVITIES_FIELDS = toField([
  '$type',
  'id',
  'timestamp',
  'targetMember',
  'targetSubMember',
  {
    authorGroup: ['icon', 'name'],
    author: IssueFields.ISSUE_USER_FIELDS,
    category: ['id'],
    target: ['id', 'created', 'usesMarkdown'],
    field: [
      '$type',
      'linkId',
      'id',
      'presentation',
      {
        customField: [
          'id',
          {
            fieldType: [
              'isMultiValue',
              'valueType'
            ]
          }
        ]
      }
    ],
    added: [
      ISSUE_PROJECT_FIELDS,

      ISSUE_ACTIVITIES_EVENT_BASE,
      IssueFields.ISSUE_COMMENTS_FIELDS,
      ISSUE_ATTACHMENT_FIELDS,

      IssueFields.ISSUE_XSHORT_FIELDS
    ],
    removed: [
      ISSUE_PROJECT_FIELDS,

      ISSUE_ACTIVITIES_EVENT_BASE,
      IssueFields.ISSUE_COMMENTS_FIELDS,

      IssueFields.ISSUE_XSHORT_FIELDS
    ]
  }
]);




export default toField([
  'cursor',
  'till',
  {
    activities: ISSUE_ACTIVITIES_FIELDS
  }
]);
