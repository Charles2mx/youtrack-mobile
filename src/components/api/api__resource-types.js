export const ResourceTypes = {
  ISSUE: 'jetbrains.charisma.persistent.Issue',
  ISSUE_COMMENT: 'jetbrains.charisma.persistent.IssueComment',

  VISIBILITY_LIMITED: 'jetbrains.charisma.persistent.visibility.LimitedVisibility',
  VISIBILITY_UNLIMITED: 'jetbrains.charisma.persistent.visibility.UnlimitedVisibility',
  VISIBILITY_GROUP: 'jetbrains.charisma.persistent.security.VisibilityGroups',

  USER: 'jetbrains.charisma.persistence.user.User',
  USER_GROUP: 'jetbrains.charisma.persistent.security.UserGroup',

  EVENT_GROUP: 'jetbrains.youtrack.event.gaprest.ActivityItemGroup',
};
