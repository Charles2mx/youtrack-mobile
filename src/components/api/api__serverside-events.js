/* @flow */
import RNEventSource from '@huston007/react-native-eventsource';
import qs from 'qs';
import log from '../../components/log/log';
import agileFields from './api__agile-fields';
import apiHelper from './api__helper';

export default class ServersideEvents {
  backendUrl: string;
  lastPing: ?Date;
  eventSource: RNEventSource;

  constructor(backendUrl: string) {
    this.backendUrl = backendUrl;
  }

  subscribeAgileBoardUpdates(ticket: string) {
    const queryString = qs.stringify(
      {
        ticket,
        fields: agileFields.liveUpdate.toString()
      },
      {encode: false}
    );

    this.eventSource = new RNEventSource(`${this.backendUrl}/api/eventSourceBus?${queryString}`);

    this.eventSource.addEventListener('open', () => log.info('SSE connection opened'));

    this.eventSource.addEventListener('error', (event) => {
      if (event?.target?.readyState !== this.eventSource?.CLOSED) {
        this.close();
      }
      return log.info('SSE connection closed', event);
    });

    this.eventSource.addEventListener('ping', () => this.lastPing = new Date());
  }

  listenTo(eventName: string, callback: any => any) {
    this.eventSource.addEventListener(eventName, event => {
      const data = event.data ? JSON.parse(event.data) : event;

      if (event.data) {
        apiHelper.patchAllRelativeAvatarUrls(data, this.backendUrl);
      }

      return callback(data);
    });
  }

  close() {
    // this.eventSource.removeAllListeners();
    this.eventSource.close();
  }
}
