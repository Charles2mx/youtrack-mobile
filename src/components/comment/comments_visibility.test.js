import React from 'react';

import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

import {ResourceTypes} from '../api/api__resource-types';
import CommentVisibility from './comment__visibility';

import type {Visibility} from '../../flow/Visibility';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';


describe('<CommentVisibility/>', () => {

  let wrapper: Object;
  let visibilityMock: Visibility;

  beforeEach(() => {
    visibilityMock = {
      $type: ResourceTypes.VISIBILITY_UNLIMITED,
      permittedUsers: [],
      permittedGroups: []
    };
    wrapper = doShallow(visibilityMock);
  });


  describe('Render', () => {
    it('should match a snapshot', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render component', () => {
      expect(findByTestId('commentVisibility')).toHaveLength(1);
      expect(findByTestId('commentVisibilityIcon')).toHaveLength(1);
    });

    it('should not render component', () => {
      wrapper = doShallow(null);

      expect(findByTestId('commentVisibility')).toHaveLength(0);
      expect(findByTestId('commentVisibilityIcon')).toHaveLength(0);
    });

    it('should colorize component', () => {
      const color = 'red';
      wrapper = doShallow(visibilityMock, color);

      expect(wrapper.contains(
        <Icon testID="commentVisibilityIcon" name="lock" size={16} color={color}/>
      )).toEqual(true);
    });
  });


  function findByTestId(testId) {
    return wrapper && wrapper.find({testID: testId});
  }

  function doShallow(visibility?: Visibility, color?: string) {
    return shallow(
      <CommentVisibility
        visibility={visibility}
        color={color}
      />
    );
  }
});
