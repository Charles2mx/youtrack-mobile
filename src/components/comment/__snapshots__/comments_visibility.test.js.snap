// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CommentVisibility/> Render should match a snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "row",
      "marginBottom": 8,
      "marginTop": 8,
    }
  }
  testID="commentVisibility"
>
  <Icon
    allowFontScaling={false}
    color="#888888"
    name="lock"
    size={16}
    testID="commentVisibilityIcon"
  />
  <Text
    style={
      Array [
        Object {
          "color": "#888888",
          "marginLeft": 5.333333333333333,
        },
        null,
      ]
    }
  />
</View>
`;
